File: packages/server/src/repowise/server/routers/workspace.py
"""/api/workspace — Workspace intelligence endpoints.

Exposes workspace metadata, cross-repo co-changes, and API contract data
through REST. All data is read from ``app.state`` (populated at server
startup from ``.repowise-workspace/`` JSON files) — no DB access needed.
"""

from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from repowise.server.deps import (
    get_cross_repo_enricher,
    get_workspace_config,
    verify_api_key,
)
from repowise.server.schemas import (
    WorkspaceCoChangeEntry,
    WorkspaceCoChangesResponse,
    WorkspaceContractEntry,
    WorkspaceContractLinkEntry,
    WorkspaceContractsResponse,
    WorkspaceContractSummary,
    WorkspaceCrossRepoSummary,
    WorkspaceRepoEntry,
    WorkspaceResponse,
)

router = APIRouter(
    prefix="/api/workspace",
    tags=["workspace"],
    dependencies=[Depends(verify_api_key)],
)


def _require_workspace(ws_config: object) -> None:
    """Raise 404 if not in workspace mode."""
    if ws_config is None:
        raise HTTPException(status_code=404, detail="Not running in workspace mode")


# ---------------------------------------------------------------------------
# GET /api/workspace
# ---------------------------------------------------------------------------


@router.get("", response_model=WorkspaceResponse)
async def get_workspace(
    request: Request,
    ws_config=Depends(get_workspace_config),
    enricher=Depends(get_cross_repo_enricher),
):
    """Workspace metadata and summary statistics.

    Returns ``is_workspace=false`` with empty data in single-repo mode —
    the web UI uses this for mode detection, so this endpoint never 404s.
    """
    if ws_config is None:
        return WorkspaceResponse(is_workspace=False)

    repo_entries = [
        WorkspaceRepoEntry(
            alias=r.alias,
            path=r.path,
            is_primary=r.is_primary,
            indexed_at=r.indexed_at,
            last_commit_at_index=r.last_commit_at_index,
        )
        for r in ws_config.repos
    ]

    cross_repo_summary = None
    contract_summary = None

    if enricher is not None:
        summary = enricher.get_cross_repo_summary()
        cross_repo_summary = WorkspaceCrossRepoSummary(**summary)
        if enricher.has_contract_data:
            cs = enricher.get_contract_summary()
            contract_summary = WorkspaceContractSummary(**cs)

    ws_root = getattr(request.app.state, "workspace_root", None)

    return WorkspaceResponse(
        is_workspace=True,
        workspace_root=ws_root,
        workspace_name=Path(ws_root).name if ws_root else None,
        repos=repo_entries,
        default_repo=ws_config.default_repo,
        cross_repo_summary=cross_repo_summary,
        contract_summary=contract_summary,
    )


# ---------------------------------------------------------------------------
# GET /api/workspace/contracts
# ---------------------------------------------------------------------------


@router.get("/contracts", response_model=WorkspaceContractsResponse)
async def get_contracts(
    ws_config=Depends(get_workspace_config),
    enricher=Depends(get_cross_repo_enricher),
    contract_type: str | None = Query(None, description="Filter: http, grpc, or topic"),
    repo: str | None = Query(None, description="Filter by repo alias"),
    role: str | None = Query(None, description="Filter: provider or consumer"),
    limit: int = Query(200, ge=1, le=1000),
    offset: int = Query(0, ge=0),
):
    """All detected contracts and matched links with optional filtering."""
    _require_workspace(ws_config)

    if enricher is None:
        return WorkspaceContractsResponse(
            contracts=[], links=[], total_contracts=0, total_links=0,
        )

    contracts = list(getattr(enricher, "_contracts", []))
    links = list(getattr(enricher, "_contract_links", []))

    # Apply filters to contracts
    if contract_type:
        contracts = [c for c in contracts if c.get("contract_type") == contract_type]
        links = [lk for lk in links if lk.get("contract_type") == contract_type]
    if repo:
        contracts = [c for c in contracts if c.get("repo") == repo]
        links = [
            lk for lk in links
            if lk.get("provider_repo") == repo or lk.get("consumer_repo") == repo
        ]
    if role:
        contracts = [c for c in contracts if c.get("role") == role]

    total_contracts = len(contracts)
    total_links = len(links)

    # Count by type
    by_type: dict[str, int] = {}
    for c in contracts:
        ct = c.get("contract_type", "unknown")
        by_type[ct] = by_type.get(ct, 0) + 1

    # Paginate contracts only (links are typically small enough)
    contracts_page = contracts[offset : offset + limit]

    return WorkspaceContractsResponse(
        contracts=[
            WorkspaceContractEntry(
                contract_id=c.get("contract_id", ""),
                contract_type=c.get("contract_type", ""),
                role=c.get("role", ""),
                repo=c.get("repo", ""),
                file_path=c.get("file_path", ""),
                symbol_name=c.get("symbol_name", ""),
                confidence=c.get("confidence", 0.0),
                service=c.get("service"),
            )
            for c in contracts_page
        ],
        links=[
            WorkspaceContractLinkEntry(
                contract_id=lk.get("contract_id", ""),
                contract_type=lk.get("contract_type", ""),
                match_type=lk.get("match_type", "exact"),
                confidence=lk.get("confidence", 0.0),
                provider_repo=lk.get("provider_repo", ""),
                provider_file=lk.get("provider_file", ""),
                provider_symbol=lk.get("provider_symbol", ""),
                consumer_repo=lk.get("consumer_repo", ""),
                consumer_file=lk.get("consumer_file", ""),
                consumer_symbol=lk.get("consumer_symbol", ""),
            )
            for lk in links
        ],
        total_contracts=total_contracts,
        total_links=total_links,
        by_type=by_type,
    )


# ---------------------------------------------------------------------------
# GET /api/workspace/co-changes
# ---------------------------------------------------------------------------


@router.get("/co-changes", response_model=WorkspaceCoChangesResponse)
async def get_co_changes(
    ws_config=Depends(get_workspace_config),
    enricher=Depends(get_cross_repo_enricher),
    repo: str | None = Query(None, description="Filter by repo alias"),
    min_strength: float = Query(0.0, ge=0.0, le=1.0),
    limit: int = Query(50, ge=1, le=500),
):
    """Cross-repo co-change pairs, optionally filtered by repo and strength."""
    _require_workspace(ws_config)

    if enricher is None:
        return WorkspaceCoChangesResponse(co_changes=[], total=0)

    co_changes = list(getattr(enricher, "_co_changes", []))

    if repo:
        co_changes = [
            cc for cc in co_changes
            if cc.get("source_repo") == repo or cc.get("target_repo") == repo
        ]
    if min_strength > 0:
        co_changes = [cc for cc in co_changes if cc.get("strength", 0) >= min_strength]

    # Sort by strength descending
    co_changes.sort(key=lambda cc: -cc.get("strength", 0))

    total = len(co_changes)
    co_changes = co_changes[:limit]

    return WorkspaceCoChangesResponse(
        co_changes=[
            WorkspaceCoChangeEntry(
                source_repo=cc.get("source_repo", ""),
                source_file=cc.get("source_file", ""),
                target_repo=cc.get("target_repo", ""),
                target_file=cc.get("target_file", ""),
                strength=cc.get("strength", 0.0),
                frequency=cc.get("frequency", 0),
                last_date=cc.get("last_date", ""),
            )
            for cc in co_changes
        ],
        total=total,
    )

File: packages/types/src/workspace.ts
/**
 * Canonical workspace types — multi-repo views (cross-repo summary, shared
 * contracts, co-changes) plus a few shared per-repo aggregates that the
 * workspace UI consumes directly.
 *
 * Canonical source: engine `WorkspaceResponse` and the per-domain rollups
 * (RepoStats, GitSummary). Downstream backends should rename via an adapter
 * to match these field names before passing to UI components.
 */

export interface RepoStats {
  file_count: number;
  symbol_count: number;
  entry_point_count: number;
  doc_coverage_pct: number;
  freshness_score: number;
  dead_export_count: number;
}

export interface WorkspaceCrossRepoSummary {
  co_change_count: number;
  package_dep_count: number;
  top_connections: Array<{ repos: string[]; edge_count: number }>;
}

export interface WorkspaceContractSummary {
  total_contracts: number;
  total_links: number;
  by_type: Record<string, number>;
}

export interface WorkspaceContractLinkEntry {
  contract_id: string;
  contract_type: string;
  match_type: string;
  confidence: number;
  provider_repo: string;
  provider_file: string;
  provider_symbol: string;
  consumer_repo: string;
  consumer_file: string;
  consumer_symbol: string;
}

export interface WorkspaceCoChangeEntry {
  source_repo: string;
  source_file: string;
  target_repo: string;
  target_file: string;
  strength: number;
  frequency: number;
  last_date: string;
}

export interface WorkspacePackageDepEntry {
  source_repo: string;
  source_manifest: string;
  target_repo: string;
  target_package: string;
  kind: string;
}

// ---------------------------------------------------------------------------
// System graph — the service-granular, typed cross-repo structure that the
// Live System Map, blast radius, the DSM, and the MCP/CLI surfaces all read.
// Mirrors `repowise.core.workspace.system_graph` (Python). Edge direction is
// uniform: `source` depends on / calls `target`.
// ---------------------------------------------------------------------------

/** Transport of a system-graph edge. `db` is reserved for a future transport. */
export type SystemEdgeKind = "http" | "grpc" | "event" | "package" | "co_change" | "db";

/** How confidently an edge was matched. Behavioral co-change edges are `inferred`. */
export type SystemEdgeMatchType = "exact" | "candidate" | "manual" | "inferred";

/** A service in the workspace (or a repo-root node when the repo is undivided). */
export interface SystemNode {
  /** Stable id: `"repo"` or `"repo::service/path"`. */
  id: string;
  /** Repo alias — the grouping attribute. */
  repo: string;
  /** Service boundary path, or null for a whole-repo node. */
  service_path: string | null;
  /** Display name (service directory basename, or repo alias). */
  name: string;
  kind: "service" | "frontend" | "worker" | "library" | "external";
  provider_count: number;
  consumer_count: number;
  contract_types: string[];
  /** Exposes provider contracts no consumer calls. */
  is_orphan_provider: boolean;
  /** Consumes contracts that never matched a provider. */
  is_orphan_consumer: boolean;
  /** Participates in no edges. */
  is_isolated: boolean;
}

/** A typed, directed relationship between two services (`source` → `target`). */
export interface SystemEdge {
  id: string;
  source: string;
  target: string;
  kind: SystemEdgeKind;
  match_type: SystemEdgeMatchType;
  confidence: number;
  /** Number of underlying contracts / co-changes / deps this edge aggregates. */
  weight: number;
  /** True for contract/package edges, false for behavioral co-change edges. */
  structural: boolean;
  /** Back-pointers to the underlying evidence, for drill-down (bounded). */
  contract_refs: string[];
}

export interface SystemGraph {
  version: number;
  generated_at: string;
  nodes: SystemNode[];
  edges: SystemEdge[];
  diagnostics: ExtractionDiagnostics;
}

// ---------------------------------------------------------------------------
// Extraction diagnostics — explains the cross-repo link count (providers /
// consumers found, unmatched-by-reason, orphan providers, weak links).
// Mirrors `repowise.core.workspace.diagnostics`.
// ---------------------------------------------------------------------------

/** Why a consumer contract never formed a cross-repo link. */
export type UnmatchedReason = "no_provider" | "internal_only" | "unlinked";

export interface RepoDiagnostics {
  repo: string;
  providers_by_type: Record<string, number>;
  consumers_by_type: Record<string, number>;
  provider_count: number;
  consumer_count: number;
}

export interface UnmatchedConsumer {
  repo: string;
  file_path: string;
  contract_id: string;
  contract_type: string;
  reason: UnmatchedReason;
}

export interface OrphanProvider {
  repo: string;
  file_path: string;
  contract_id: string;
  contract_type: string;
}

export interface ExtractionDiagnostics {
  total_providers: number;
  total_consumers: number;
  total_links: number;
  weak_link_count: number;
  repo_breakdown: RepoDiagnostics[];
  unmatched_consumers: UnmatchedConsumer[];
  unmatched_by_reason: Record<string, number>;
  orphan_providers: OrphanProvider[];
}

// ---------------------------------------------------------------------------
// Cross-repo blast radius — reachability over the system graph. "If I change
// this service, what downstream services and repos break?" Mirrors the
// single-repo blast-radius vocabulary (`blast-radius.ts`): impacted items carry
// an impact `score` and a `distance`. Mirrors `repowise.core.workspace.blast_radius`.
// ---------------------------------------------------------------------------

/** A service reachable from the change, with its ranked impact. */
export interface ImpactedNode {
  /** System-graph node id (`"repo"` or `"repo::service/path"`). */
  id: string;
  repo: string;
  name: string;
  kind: "service" | "frontend" | "worker" | "library" | "external";
  /** Hops from the nearest changed node (1 = a direct dependent). */
  distance: number;
  /** 0-1 ranked impact, with distance decay and behavioral weighting baked in. */
  score: number;
  /** Reachable via an all-structural path (a real dependency, not co-change). */
  structural: boolean;
  /** Distinct edge kinds that carried impact into this node. */
  edge_kinds: SystemEdgeKind[];
}

export interface CrossRepoBlastRadius {
  /** Resolved node ids the traversal started from. */
  targets: string[];
  /** Distinct repos of the targets. */
  target_repos: string[];
  /** Ranked impact set (strongest first), capped server-side. */
  impacted: ImpactedNode[];
  /** Distinct repos in the impact set, excluding the target repos. */
  impacted_repos: string[];
  /** Impacted nodes reachable via a real dependency. */
  structural_count: number;
  /** Impacted nodes reachable only via co-change correlation. */
  behavioral_count: number;
  max_distance: number;
  /** True count before the server-side cap. */
  total_impacted: number;
  /** Target strings that matched no node or repo. */
  unresolved_targets: string[];
}

// ---------------------------------------------------------------------------
// Contract schema — the optional request/response field shape a contract
// carries when a parser can recover it (proto message fields today; OpenAPI
// when present). Drives schema-level breaking-change detection.
// Mirrors `repowise.core.workspace.contract_schema`.
// ---------------------------------------------------------------------------

/** One field in a request or response shape. `number` is the proto wire tag. */
export interface SchemaField {
  name: string;
  type: string;
  required?: boolean;
  number?: number | null;
  repeated?: boolean;
}

export interface ContractSchema {
  /** Which parser produced the shape (`"proto"` / `"openapi"`). */
  source: string;
  request_fields: SchemaField[];
  response_fields: SchemaField[];
}

// ---------------------------------------------------------------------------
// Breaking-change guard — provider contract changes that break consumers across
// repos, computed by diffing the current contracts against the previously
// indexed set. Mirrors `repowise.core.workspace.breaking_change`.
// ---------------------------------------------------------------------------

/** How a breaking change ranks. `breaking` = wire-incompatible; `warning` = source risk. */
export type BreakingChangeSeverity = "breaking" | "warning";

/** A consumer endangered by a provider's breaking change (from a matched link). */
export interface BreakingChangeConsumer {
  repo: string;
  service: string | null;
  /** System-graph node id of the consumer (for map badging). */
  node_id: string;
  /** The exact consumer file that calls the changed contract. */
  file: string;
  symbol: string;
  match_type: string;
  confidence: number;
}

export interface BreakingChange {
  /** Rule key: `removed_endpoint` | `removed_field` | `field_type_changed` | ... */
  kind: string;
  severity: BreakingChangeSeverity;
  contract_id: string;
  contract_type: string;
  provider_repo: string;
  provider_file: string;
  provider_symbol: string;
  provider_service: string | null;
  /** System-graph node id of the changed provider. */
  provider_node_id: string;
  /** Human-readable one-liner. */
  detail: string;
  field_name?: string | null;
  old_value?: string | null;
  new_value?: string | null;
  impacted_consumers: BreakingChangeConsumer[];
}

export interface BreakingChangeReport {
  version: number;
  generated_at: string;
  changes: BreakingChange[];
  total: number;
  breaking_count: number;
  warning_count: number;
  /** Distinct repos with an endangered consumer. */
  impacted_repos: string[];
  /** Distinct system-graph node ids with an endangered consumer. */
  impacted_services: string[];
  total_impacted_consumers: number;
}
